native/src/main/java/io/ballerina/stdlib/data/jsondata/json/JsonTraverse.java

```diff
@@ -0,0 +1,301 @@
+                        }
+                    }
+                    return DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, type, PredefinedTypes.TYPE_ANYDATA);
+                case TypeTags.JSON_TAG:
+                case TypeTags.ANYDATA_TAG:
+                    return json;
+                case TypeTags.MAP_TAG:
+                    MapType mapType = (MapType) referredType;
+                    fieldHierarchy.push(new HashMap<>());
+                    restType.push(mapType.getConstrainedType());
+                    return traverseMapJsonOrArrayJson(json, ValueCreator.createMapValue(mapType), referredType);
+                default:
+                    return DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, type, PredefinedTypes.TYPE_ANYDATA);
+            }
+        }
+
+        private Object traverseMapJsonOrArrayJson(Object json, Object currentJsonNode, Type type) {
+            if (json instanceof BMap map) {
+                return traverseMapValue(map, currentJsonNode);
+            } else if (json instanceof BArray) {
+                return traverseArrayValue(json, currentJsonNode);
+            } else {
+                // JSON value not compatible with map or array.
+                if (type.getTag() == TypeTags.RECORD_TYPE_TAG) {
+                    this.fieldHierarchy.pop();
+                    this.restType.pop();
+                }
+
+                if (fieldNames.isEmpty()) {
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INCOMPATIBLE_TYPE, type, json);
+                }
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE_FOR_FIELD, getCurrentFieldPath());
+            }
+        }
+
+        private Object traverseMapValue(BMap<BString, Object> map, Object currentJsonNode) {
+            for (BString key : map.getKeys()) {
+                currentField = fieldHierarchy.peek().remove(key.toString());
+                if (currentField == null) {
+                    // Add to the rest field
+                    if (restType.peek() != null) {
+                        Type restFieldType = TypeUtils.getReferredType(restType.peek());
+                        addRestField(restFieldType, key, map.get(key), currentJsonNode);
+                    }
+                    continue;
+                }
+
+                String fieldName = currentField.getFieldName();
+                fieldNames.push(fieldName);
+                Type currentFieldType = TypeUtils.getReferredType(currentField.getFieldType());
+                int currentFieldTypeTag = currentFieldType.getTag();
+                Object mapValue = map.get(key);
+
+                switch (currentFieldTypeTag) {
+                    case TypeTags.NULL_TAG:
+                    case TypeTags.BOOLEAN_TAG:
+                    case TypeTags.INT_TAG:
+                    case TypeTags.FLOAT_TAG:
+                    case TypeTags.DECIMAL_TAG:
+                    case TypeTags.STRING_TAG:
```

Comment:
if there is a field value of type `boolean?` this will not work as expected ryt?

Reply:
```
import ballerina/data.jsondata;
import ballerina/io;

public function main() returns error? {
    json jsonVal = {
        isTrue: true,
        isFalse: false,
        name: "John"
    };

    record {|
        boolean isTrue;
        boolean isFalse;
        string name;
    |} rec = check jsondata:fromJsonWithType(jsonVal);
    io:println(rec);
}
```
I tried this sample, it worked as expected. Boolean case is covered. Right?  May I know the exact use case you are referring to?